// react-smart-image/dist/client.js
const Promise = require('promise');
const React = require('react');
const _ = require('underscore');

const ImgMixin = {
  _isMounted: false,
  getInitialState: function() {
    return {
      src: this.props.src,
      src_fallback: this.srcFallback(),
      opacity: 0,
      status: 'loading'
    };
  },
  srcFallback: function() {
    if (this.props.srcFallback) return this.props.srcFallback;
    const style = this.props.style || {};
    const width = this.props.width || style.width || 100;
    const height = this.props.height || style.height || 100;
    return '//lorempixel.com/' + width + '/' + height + '/cats?v=' + _.random(100);
  },
  loadImage: function(src) {
    return new Promise(function(resolve, reject) {
      const img = document.createElement('IMG');
      img.src = src;
      img.onload = resolve;
      img.onerror = reject;
    });
  },
  renderImage: function() {
    this._isMounted && this.setState({ status: 'loading' });
    this.loadImage(this.props.src)
      .then(function() {
        this._isMounted && this.setState({ src: this.props.src, opacity: 1, status: 'loaded' });
      }.bind(this))
      .catch(function() {
        this.loadImage(this.state.src_fallback)
          .then(function() {
            this._isMounted && this.setState({ src: this.state.src_fallback });
            setTimeout(function() {
              this._isMounted && this.setState({ opacity: 1, status: 'error' });
            }.bind(this), 100);
          }.bind(this))
          .catch(function(){});
      }.bind(this));
  },
  componentDidMount: function() {
    this._isMounted = true;
    this.renderImage();
  },
  componentWillUpdate: function(props, state) {
    if (this.props.src !== props.src) {
      this._isMounted && this.setState({ src: null, opacity: 0, status: 'loading' });
    }
  },
  componentDidUpdate: function(props, state) {
    if (this.props.src !== props.src) {
      this.renderImage();
    }
  },
  componentWillUnmount: function(props, state) {
    this._isMounted = false;
  }
};

const Img = React.createClass({displayName: "Img",
  mixins: [ImgMixin],
  attrs: function() {
    return _.extend(_.omit(this.props, 'src') || {}, {
      src: this.state.src,
      className: [this.props.className, 'img-' + this.state.status].join(' '),
      style: _.extend(this.props.style || {}, {
        opacity: this.state.opacity,
        transition: 'opacity 0.6s ease'
      })
    });
  },
  render: function() {
    return React.DOM.img(this.attrs());
  }
});

const BgMixin = {
  attrs: function() {
    return _.extend(_.omit(this.props, 'src') || {}, {
      className: [this.props.className, 'div-bg-' + this.state.status].join(' '),
      style: _.extend(this.props.style || {}, {
        opacity: this.state.opacity,
        transition: 'opacity 0.6s ease',
        backgroundSize: 'cover',
        backgroundPosition: 'center',
        backgroundRepeat: 'no-repeat',
        backgroundImage: 'url(' + this.state.src + ')'
      })
    });
  }
};

const DivBg = React.createClass({displayName: "DivBg",
  mixins: [ImgMixin, BgMixin],
  render: function() {
    return (React.createElement('div', this.attrs(), this.props.children));
  }
});

const ABg = React.createClass({displayName: "ABg",
  mixins: [ImgMixin, BgMixin],
  render: function() {
    return (React.createElement('a', this.attrs(), this.props.children));
  }
});

module.exports = { Img: Img, DivBg: DivBg, ABg: ABg };
